### *“If you’re going through hell, keep going.”* 
*-Winston Churchill, Prime Minister of the UK* 

*I know this time of the semester is ROUGH but the end is in sight, so keep moving forward!*

*Relaxing times are near! -Dr. R*

----

YOUR NAME HERE: Daniel DeLuca

----

# Lab P: Applying Functions to Columns

We've learned a lot over the last few weeks.  With your tools so far, you can:

1. Load a dataset from the web;
2. Work with (extract, add, drop, relabel) columns from the dataset;
3. Filter and sort it according to certain criteria;
4. Perform arithmetic on columns of numbers;
5. Group rows by columns of categories, counting the number of rows in each category;
6. Make bar charts, line graphs, scatter plots, and histograms.

These tools are fairly powerful, but they're not quite enough for all the analysis and data we'll do.  Today we'll learn a tool that dramatically expands this toolbox: the table method `apply`.  

In [175]:
# Run this cell to set up the notebook
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
plt.style.use('fivethirtyeight')

## 1. Functions and CEO Incomes

Let's start with a real data analysis task.  We'll look at the 2015 compensation of CEOs at the 100 largest companies in California.  The data were compiled for a Los Angeles Times analysis [here](http://spreadsheets.latimes.com/california-ceo-compensation/), and ultimately came from [filings](https://www.sec.gov/answers/proxyhtf.htm) mandated by the SEC from all publicly-traded companies.  Two companies have two CEOs, so there are 102 CEOs in the dataset.

We've copied the data in raw form from the LA Times page into a file called `raw_compensation.csv`.  (The page notes that all dollar amounts are in millions of dollars.)

In [74]:
raw_compensation = pd.read_csv('raw_compensation.csv')
raw_compensation

Rank                Name                 Company (Headquarters)  \
0       1       Mark V. Hurd*                  Oracle (Redwood City)   
1       2      Safra A. Catz*                  Oracle (Redwood City)   
2       3      Robert A. Iger                  Walt Disney (Burbank)   
3       4    Marissa A. Mayer                     Yahoo! (Sunnyvale)   
4       5        Marc Benioff         salesforce.com (San Francisco)   
..    ...                 ...                                    ...   
97     98  Gregory N. Roberts  A-Mark Precious Metals (Santa Monica)   
98     99     Jose L. Laparte                 PriceSmart (San Diego)   
99    100         Jack Dorsey                Twitter (San Francisco)   
100   101        Elon R. Musk               Tesla Motors (Palo Alto)   
101   102    Lawrence Page***               Alphabet (Mountain View)   

    Total Pay            % Change Cash Pay Equity Pay Other Pay  \
0     $53.25   (No previous year)   $0.95     $52.27     $0.02    
1     $53.24   (No previous year)   $0.95     $52.27     $0.02    
2     $44.91                  -3%  $24.89     $17.28     $2.74    
3     $35.98                 -15%   $1.00     $34.43     $0.55    
4     $33.36                 -16%   $4.65     $27.26     $1.45    
..        ...                 ...      ...        ...       ...   
97     $0.94                 -11%   $0.93      $0.00     $0.02    
98     $0.81                  -7%   $0.71      $0.00     $0.10    
99     $0.07   (No previous year)   $0.00      $0.00     $0.07    
100    $0.04                   6%   $0.04      $0.00     $0.00    
101    $0.00                   0%   $0.00      $0.00     $0.00    

     Ratio of CEO pay to average industry worker pay  
0                                          362.00000  
1                                          362.00000  
2                                          477.00000  
3                                          342.00000  
4                                          338.00000  
..                                               ...  
97                                          17.00000  
98                                          32.00000  
99                                           0.63000  
100                                          0.46000  
101                                          0.00001  

[102 rows x 9 columns]

---

# **Question 1.1**

When I first loaded this dataset, I tried to compute the average of the CEOs' pay like this:

    raw_compensation["Total Pay"].mean()

Explain why that didn't work.  *Hint:* Try looking at some of the values in the "Total Pay" column.

groupby should be used instead of just taking the mean of the columns because the columns have a dollar sign in it

In [75]:
raw_compensation.groupby('Total Pay').mean()

Rank  Ratio of CEO pay to average industry worker pay
Total Pay                                                        
$0.00      102.0                                          0.00001
$0.04      101.0                                          0.46000
$0.07      100.0                                          0.63000
$0.81       99.0                                         32.00000
$0.94       98.0                                         17.00000
...          ...                                              ...
$9.42       54.0                                         82.00000
$9.74       53.0                                        139.00000
$9.86       52.0                                        324.00000
$9.96       51.0                                        280.00000
$9.97       50.0                                         72.00000

[100 rows x 2 columns]

---

# **Question 1.2**
The answer to the previous question was that the total pay column doesn't have numbers in it - It has strings!  You can see this because of the dollar sign.

Extract the first value in the "Total Pay" column.  It's Mark Hurd's pay in 2015, in *millions* of dollars.  Call it `mark_hurd_pay_string`.

In [76]:
mark_hurd_pay_string = raw_compensation['Total Pay'].iloc[0]
mark_hurd_pay_string

'$53.25 '

---

# **Question 1.3**

Convert `mark_hurd_pay_string` to a number of *dollars*.  The string method `strip` will be useful for removing the dollar sign; it removes a specified character from the start or end of a string.  For example, the value of `"100%".strip("%")` is the string `"100"`.  You'll also need the function `float`, which converts a string that looks like a number to an actual number.  Last, remember that the answer should be in dollars, not millions of dollars.

In [77]:
mark_hurd_pay_string = (mark_hurd_pay_string.strip('$')

SyntaxError: unexpected EOF while parsing (Temp/ipykernel_46328/2564687472.py, line 1)

In [ ]:
mark_hurd_pay_string = float(mark_hurd_pay_string)
mark_hurd_pay_string 

---
To compute the average pay, we need to do this for every CEO.  But that looks like it would involve copying this code 102 times.

This is where functions come in.  First, we'll define our own function that packages together the code we wrote to convert a pay string to a pay number.  This has its own benefits.  Later in this lab we'll see a bigger payoff: we can call that function on every pay string in the dataset at once.

---
# **Question 1.4**

Below we've written code that defines a function that converts pay strings to pay numbers, just like your code above.  

But it has a small error, and if you read carefully, you can see what it is.

Even if  you don't known what all the other stuff in the cell means, you have the ability to fix the error, so...  

...Correct the problem.

In [ ]:
def convert_pay_string_to_number(pay_string):
    """Converts a pay string like '$100' (in millions) 
    to a number of dollars (so NOT in millions)."""
    
    return float(pay_string.strip("$")) *1000000  #What should this return?

---

Running that cell doesn't convert any particular pay string.

Rather, think of it as defining a *recipe* for converting a pay string to a number.  Writing down a recipe for cake doesn't give you a cake.  You have to gather the ingredients and get a chef to execute the instructions in the recipe to get a cake.  Similarly, no pay string is converted to a number until we *call* our function on a particular pay string (which tells Python, our lightning-fast chef, to execute it).

We can call our function just like we call the built-in functions we've seen.  (Almost all of those functions are defined in this way, in fact!)  It takes one argument, a string, and it returns a number.

In [ ]:
convert_pay_string_to_number(mark_hurd_pay_string)

#### HINT:  You should see 53250000 printed out if you truly corrected the issue in Question 1.4.

---

# **Question 1.5**

We can also compute Safra Catz's pay in the same way! Complete the code below to make that happen.

In [ ]:
catz_row = raw_compensation[raw_compensation["Rank"] == 2]
catz_row

In [ ]:
convert_pay_string_to_number(catz_row["Total Pay"].iloc[0])

---

What have we gained?  Well, without the function, we'd have to copy that `10**6 * float(pay_string.strip("$"))` stuff each time we wanted to convert a pay string.  Now we just call a function whose name says exactly what it's doing.

We'd still have to call the function 102 times to convert all the salaries, which we'll fix next.

But for now, let's write some more functions.

## 2. Defining functions
##### In Which We Write a Lot of Recipes

Let's write a very simple function that converts a proportion to a percentage by multiplying it by 100.  For example, the value of `to_percentage(.5)` should be the number 50.  (No percent sign.)

A function definition has a few parts.

##### `def`
It always starts with `def` (short for **def**ine):

    def

##### Name
Next comes the name of the function.  Let's call our function `to_percentage`.
    
    def to_percentage

##### Signature
Next comes something called the *signature* of the function.  This tells Python how many arguments your function should have, and what names you'll use to refer to those arguments in the function's code.  `to_percentage` should take one argument, and we'll call that argument `proportion` since it should be a proportion.

    def to_percentage(proportion)

We put a colon after the signature to tell Python it's over.

    def to_percentage(proportion):

##### Documentation
Functions can do complicated things, so you should write an explanation of what your function does.  For small functions, this is less important, but it's a good habit to learn from the start.  Conventionally, Python functions are documented by writing a triple-quoted string:

    def to_percentage(proportion):
        """Converts a proportion to a percentage."""
    
    
##### Body
Now we start writing code that runs when the function is called.  This is called the *body* of the function.  We can write anything we could write anywhere else.  First let's give a name to the number we multiply a proportion by to get a percentage.

    def to_percentage(proportion):
        """Converts a proportion to a percentage."""
        factor = 100

##### `return`
The special instruction `return` in a function's body tells Python to make the value of the function call equal to whatever comes right after `return`.  We want the value of `to_percentage(.5)` to be the proportion .5 times the factor 100, so we write:

    def to_percentage(proportion):
        """Converts a proportion to a percentage."""
        factor = 100
        return proportion * factor
        
        
**Notice that the code in the function must be indented and that the function ends in a return, which is typical in this class.**

---

# **Question 2.1**

Define `to_percentage` in the cell below.  Call your function to convert the proportion .2 to a percentage.  Name that percentage `twenty_percent`.

In [ ]:
def to_percentage(proportion):
    """converts a propotion to a percentage"""
    factor = 100
    return proportion * factor 

twenty_percent = to_percentage(.2)
twenty_percent

---

Like the built-in functions, you can use named values as arguments to your function.

# **Question 2.2**

Use `to_percentage` again to convert the proportion named `a_proportion` (defined below) to a percentage called `a_percentage`.

*Note:* You don't need to define `to_percentage` again!  Just like other named things, functions stick around after you define them.

In [ ]:
a_proportion = 2**(.5) / 2
a_percentage = to_percentage(a_proportion)
a_percentage

---

### Here's something important about functions: 

Each time a function is called, it creates its own "space" for names that's separate from the main space where you normally define names.  (Exception: all the names from the main space get copied into it.)  So even though you defined `factor = 100` inside `to_percentage` above and then called `to_percentage`, you can't refer to `factor` anywhere except inside the body of `to_percentage`:

In [ ]:
# You should see an error when you run this.  (If you don't, you might
# have defined factor somewhere above.)
factor

As we've seen with the built-in functions, functions can also take strings (or arrays, or tables) as arguments, and they can return those things, too.

---
# **Question 2.3**

Define a function called `disemvowel`.  It should take a single string as its argument.  (You can call that argument whatever you want.)  It should return a copy of that string, but with all the characters that are vowels removed.  (In English, the vowels are the characters "a", "e", "i", "o", and "u".)

*Hint:* 

* To remove all the "a"s from a string, you can use `that_string.replace("a", "")`.  
* And you can call `replace` multiple times.

In [ ]:
def disemvowel(a_string):
    """removes vowels from a string"""
    return a_string.replace("a","").replace("e","").replace("i","").replace("o","").replace("u","")

# An example call to your function.  (It's often helpful to run
# an example call from time to time while you're writing a function,
# to see how it currently works.)
disemvowel("Can you read this without vowels?")

---

##### Calls on calls on calls
Just as you write a series of lines to build up a complex computation, it's useful to define a series of small functions that build on each other.  Since you can write any code inside a function's body, you can call other functions you've written.

This is like a recipe for cake telling you to follow another recipe to make the frosting, and another to make the sprinkles.  This makes the cake recipe shorter and clearer, and it avoids having a bunch of duplicated frosting recipes.  It's a foundation of productive programming.

For example, suppose you want to count the number of characters *that aren't vowels* in a piece of text.  One way to do that is this to remove all the vowels and count the size of the remaining string.

---

# **Question 2.4**

Write a function called `num_non_vowels`.  It should take a string as its argument and return a number.  The number should be the number of characters in the argument string that aren't vowels.

*Hint:* Recall that the function `len` takes a string as its argument and returns the number of characters in it.

In [ ]:
def num_non_vowels(a_string):
    """The number of characters in a string, minus the vowels."""
    factor =  a_string.replace("a","").replace("e","").replace("i","").replace("o","").replace("u","")
    return len(factor) 
num_non_vowels("I love CS-260 DataBases and Data Visualization")

---

Functions can also encapsulate code that *does things* rather than just computing values.  For example, if you call `print` inside a function, and then call that function, something will get printed.

The `movies_by_year` dataset in the textbook has information about movie sales a few years ago.  Suppose you'd like to display the year with the 5th-highest total gross movie sales, printed in a human-readable way.  You might do this:

In [ ]:
movies_by_year = pd.read_csv("movies_by_year.csv")
rank = 5
fifth_from_top_movie_year = movies_by_year.sort_values("Total Gross", ascending=False)["Year"].iloc[rank-1]
print("Year number", rank, "for total gross movie sales was:", fifth_from_top_movie_year)

After writing this, you realize you also wanted to print out the 2nd and 3rd-highest years.  Instead of copying your code, you decide to put it in a function.  Since the rank varies, you make that an argument to your function.

---
# **Question 2.5**

Write a function called `print_kth_top_movie_year`.  It should take a single argument, the rank of the year (like 2, 3, or 5 in the above examples).  It should print out a message like the one above.  It shouldn't have a `return` statement because it's just printing and not calculating anything.

In [ ]:
rank = 2
def print_kth_top_movie_year(k):
    movie_rank = movies_by_year.sort_values("Total Gross", ascending = False)["Year"].iloc[rank-1]
    print("Year number", rank, "for total gross movie sales was:" , movie_rank) 
   

# Example calls to your function:
print_kth_top_movie_year(2)
print_kth_top_movie_year(3)

---

## 3. `apply`ing functions
##### In Which Python Bakes 102 Cakes
You'll get more practice writing functions, but let's move on.  

Defining a function is a lot like giving a name to a value with `=`.  In fact, a function is a value just like the number 1 or the text "the"!

For example, we can make a new name for the built-in function `max` if we want:

In [ ]:
our_name_for_max = max
our_name_for_max(2, 6)

The old name for `max` is still around:

In [ ]:
max(2, 6)

Try just writing `max` or `our_name_for_max` (or the name of any other function) in a cell, and run that cell.  Python will print out a (very brief) description of the function.

In [ ]:
max

Why is this useful?  Since functions are just values, it's possible to pass them as arguments to other functions.  Here's a simple but not-so-practical example: we can make an array of functions.

In [ ]:
np.array([max, np.mean, print])

---

# **Question 3.1**

Make an array containing any 3 other functions you've seen.  Call it `some_functions`.

In [ ]:
some_functions = np.array([min , float , len])
some_functions

Working with functions as values can lead to some funny-looking code.  For example, see if you can figure out why this works:

In [ ]:
np.array([max, np.mean, print])[0](4, -2, 7)

**Here's a simpler example that's actually useful: the DataFrame method `apply`.**

`apply` calls a function many times, once on *each* element in a column of a table.  It produces an array of the results.  Here we use `apply` to convert every CEO's pay to a number, using the function you defined:

In [78]:
raw_compensation['Total Pay'].apply(convert_pay_string_to_number)

0      53250000.0
1      53240000.0
2      44910000.0
3      35980000.0
4      33360000.0
          ...    
97       940000.0
98       810000.0
99        70000.0
100       40000.0
101           0.0
Name: Total Pay, Length: 102, dtype: float64

Here's an illustration of what that did:

<img src="https://raw.githubusercontent.com/kathleen-ryan-DeSales/CS250/main/apply.png"/>

Note that we didn't write something like `convert_pay_string_to_number()` or `convert_pay_string_to_number("Total Pay")`.  The job of `apply` is to call the function we give it, so instead of calling `convert_pay_string_to_number` ourselves, we just write its name as an argument to `apply`.

---

# **Question 3.2**

Make a copy of the ``raw_compensation`` table and call it ``compensation``. In the ``compensation`` table, make a column called "Total Pay (\$)".  It should be the result of applying `convert_pay_string_to_number` to the "Total Pay" column, as we did above.  

In [80]:
compensation = raw_compensation 
compensation["Total Pay $"] = compensation['Total Pay'].apply(convert_pay_string_to_number)
compensation

Rank                Name                 Company (Headquarters)  \
0       1       Mark V. Hurd*                  Oracle (Redwood City)   
1       2      Safra A. Catz*                  Oracle (Redwood City)   
2       3      Robert A. Iger                  Walt Disney (Burbank)   
3       4    Marissa A. Mayer                     Yahoo! (Sunnyvale)   
4       5        Marc Benioff         salesforce.com (San Francisco)   
..    ...                 ...                                    ...   
97     98  Gregory N. Roberts  A-Mark Precious Metals (Santa Monica)   
98     99     Jose L. Laparte                 PriceSmart (San Diego)   
99    100         Jack Dorsey                Twitter (San Francisco)   
100   101        Elon R. Musk               Tesla Motors (Palo Alto)   
101   102    Lawrence Page***               Alphabet (Mountain View)   

    Total Pay            % Change Cash Pay Equity Pay Other Pay  \
0     $53.25   (No previous year)   $0.95     $52.27     $0.02    
1     $53.24   (No previous year)   $0.95     $52.27     $0.02    
2     $44.91                  -3%  $24.89     $17.28     $2.74    
3     $35.98                 -15%   $1.00     $34.43     $0.55    
4     $33.36                 -16%   $4.65     $27.26     $1.45    
..        ...                 ...      ...        ...       ...   
97     $0.94                 -11%   $0.93      $0.00     $0.02    
98     $0.81                  -7%   $0.71      $0.00     $0.10    
99     $0.07   (No previous year)   $0.00      $0.00     $0.07    
100    $0.04                   6%   $0.04      $0.00     $0.00    
101    $0.00                   0%   $0.00      $0.00     $0.00    

     Ratio of CEO pay to average industry worker pay  Total Pay $  
0                                          362.00000   53250000.0  
1                                          362.00000   53240000.0  
2                                          477.00000   44910000.0  
3                                          342.00000   35980000.0  
4                                          338.00000   33360000.0  
..                                               ...          ...  
97                                          17.00000     940000.0  
98                                          32.00000     810000.0  
99                                           0.63000      70000.0  
100                                          0.46000      40000.0  
101                                          0.00001          0.0  

[102 rows x 10 columns]

- - - - - - - - - - - - - - - - - - - - - - - - - - -

Now that we have the pay in numbers, we can compute things about them.

# **Question 3.3**

Compute the average total pay of the CEOs in the dataset.

In [81]:
average_total_pay = compensation["Total Pay $"].mean()
average_total_pay

11445294.11764706

---

# **Question 3.4**

Companies pay executives in a variety of ways: directly in cash; by granting stock or other "equity" in the company; or with ancillary benefits (like private jets).  Compute the proportion of each CEO's pay that was cash.  Add a column to the ```compensation``` table called "% Cash" to store this information.

In [83]:
compensation["Cash $"] = compensation['Cash Pay'].apply(convert_pay_string_to_number)
compensation

Rank                Name                 Company (Headquarters)  \
0       1       Mark V. Hurd*                  Oracle (Redwood City)   
1       2      Safra A. Catz*                  Oracle (Redwood City)   
2       3      Robert A. Iger                  Walt Disney (Burbank)   
3       4    Marissa A. Mayer                     Yahoo! (Sunnyvale)   
4       5        Marc Benioff         salesforce.com (San Francisco)   
..    ...                 ...                                    ...   
97     98  Gregory N. Roberts  A-Mark Precious Metals (Santa Monica)   
98     99     Jose L. Laparte                 PriceSmart (San Diego)   
99    100         Jack Dorsey                Twitter (San Francisco)   
100   101        Elon R. Musk               Tesla Motors (Palo Alto)   
101   102    Lawrence Page***               Alphabet (Mountain View)   

    Total Pay            % Change Cash Pay Equity Pay Other Pay  \
0     $53.25   (No previous year)   $0.95     $52.27     $0.02    
1     $53.24   (No previous year)   $0.95     $52.27     $0.02    
2     $44.91                  -3%  $24.89     $17.28     $2.74    
3     $35.98                 -15%   $1.00     $34.43     $0.55    
4     $33.36                 -16%   $4.65     $27.26     $1.45    
..        ...                 ...      ...        ...       ...   
97     $0.94                 -11%   $0.93      $0.00     $0.02    
98     $0.81                  -7%   $0.71      $0.00     $0.10    
99     $0.07   (No previous year)   $0.00      $0.00     $0.07    
100    $0.04                   6%   $0.04      $0.00     $0.00    
101    $0.00                   0%   $0.00      $0.00     $0.00    

     Ratio of CEO pay to average industry worker pay  Total Pay $      Cash $  
0                                          362.00000   53250000.0    950000.0  
1                                          362.00000   53240000.0    950000.0  
2                                          477.00000   44910000.0  24890000.0  
3                                          342.00000   35980000.0   1000000.0  
4                                          338.00000   33360000.0   4650000.0  
..                                               ...          ...         ...  
97                                          17.00000     940000.0    930000.0  
98                                          32.00000     810000.0    710000.0  
99                                           0.63000      70000.0         0.0  
100                                          0.46000      40000.0     40000.0  
101                                          0.00001          0.0         0.0  

[102 rows x 11 columns]

In [3]:

cash_proportion = compensation["Cash $"]/compensation["Total Pay $"]
compensation["cash_proportion"] = cash_proportion
compensation

NameError: name 'compensation' is not defined

Check out the "% Change" column in `compensation`.  It shows the percentage increase in the CEO's pay from the previous year.  For CEOs with no previous year on record, it instead says "(No previous year)".  The values in this column are *strings*, not numbers, so like the "Total Pay" column, it's not usable without a bit of extra work.

Given your current pay and the percentage increase from the previous year, you can compute your previous year's pay.  For example, if your pay is 115 dollars this year, and that's an increase of 50% from the previous year, then your previous year's pay was $$\frac{115}{1 + \frac{50}{100}}$$ so around 76.66 dollars.

---

# **Question 3.5** look at picture on blackboard

Step 1:  Create a new table called `with_previous_compensation`.  It should be a copy of `compensation`, but with the "(No previous year)" CEOs filtered out, and with an extra column called "2014 Total Pay ($)".  That column should have each CEO's pay in 2014.

In [2]:
compensation['% Cash'] = (compensation['Cash Pay'].apply(convert_pay_string_to_number) / compensation['Total Pay $']) *100
compensation

NameError: name 'compensation' is not defined

In [1]:

with_previous_compensation = compensation[compensation['% Change'] != "(No previous year)"].copy()

NameError: name 'compensation' is not defined

Step 2:  Use the equation given above to create a column called ``Previous Pay ($)`` that shows what each CEO made last year.  If you do this right, then you will see that Marissa A. Mayer made $42,329,412 in the previous year.

* *Hint 1:* This question takes several steps, but each one is still something you've seen before.  Take it one step at a time, using as many lines as you need.  You can print out your results after each step to make sure you're on the right track.

* *Hint 2:* You'll need to define a function.  You can do that just above your other code.

* *Hint 3:*  You will likely see the following error while doing this.  This error comes up often.  It means that you are performing calculations on only some rows of the original DataFrame and Python is worried about the other rows being left out.  You see, when filtering out the "(No previous year)" rows, Python does not give you a true copy of the original DataFrame, but just a filtered down view into it.  You want a true copy to do this computation.  To fix this, just go back to the Part 1 cell and get a true copy of the DataFrame by adding ```.copy()``` to the end of the line where you have  ```with_previous_compensation = compensation[...]```.

    - ```A value is trying to be set on a copy of a slice from a DataFrame.```

In [162]:

def convert_pay_percent_to_number(pay_string):
    return float(pay_string.strip('%'))/100



In [165]:
percentage_change = with_previous_compensation['% Change'].apply(convert_pay_percent_to_number)
percentage_change

2     -0.03
3     -0.15
4     -0.16
5     -0.04
6     -0.15
       ... 
96    -0.66
97    -0.11
98    -0.07
100    0.06
101    0.00
Name: % Change, Length: 81, dtype: float64

In [167]:
with_previous_compensation['Previous Pay $'] = (with_previous_compensation['Total Pay $']) / (1 + percentage_change)
with_previous_compensation

Rank                Name                 Company (Headquarters)  \
2       3      Robert A. Iger                  Walt Disney (Burbank)   
3       4    Marissa A. Mayer                     Yahoo! (Sunnyvale)   
4       5        Marc Benioff         salesforce.com (San Francisco)   
5       6  John H. Hammergren               McKesson (San Francisco)   
6       7      John S. Watson                    Chevron (San Ramon)   
..    ...                 ...                                    ...   
96     97       David G. Hirz        Smartæ& Final Stores (Commerce)   
97     98  Gregory N. Roberts  A-Mark Precious Metals (Santa Monica)   
98     99     Jose L. Laparte                 PriceSmart (San Diego)   
100   101        Elon R. Musk               Tesla Motors (Palo Alto)   
101   102    Lawrence Page***               Alphabet (Mountain View)   

    Total Pay % Change Cash Pay Equity Pay Other Pay  \
2     $44.91       -3%  $24.89     $17.28     $2.74    
3     $35.98      -15%   $1.00     $34.43     $0.55    
4     $33.36      -16%   $4.65     $27.26     $1.45    
5     $24.84       -4%  $12.10     $12.37     $0.37    
6     $22.04      -15%   $4.31     $14.68     $3.05    
..        ...      ...      ...        ...       ...   
96     $1.53      -66%   $1.50      $0.00     $0.04    
97     $0.94      -11%   $0.93      $0.00     $0.02    
98     $0.81       -7%   $0.71      $0.00     $0.10    
100    $0.04        6%   $0.04      $0.00     $0.00    
101    $0.00        0%   $0.00      $0.00     $0.00    

     Ratio of CEO pay to average industry worker pay  Total Pay $      Cash $  \
2                                          477.00000   44910000.0  24890000.0   
3                                          342.00000   35980000.0   1000000.0   
4                                          338.00000   33360000.0   4650000.0   
5                                          222.00000   24840000.0  12100000.0   
6                                          183.00000   22040000.0   4310000.0   
..                                               ...          ...         ...   
96                                          26.00000    1530000.0   1500000.0   
97                                          17.00000     940000.0    930000.0   
98                                          32.00000     810000.0    710000.0   
100                                          0.46000      40000.0     40000.0   
101                                          0.00001          0.0         0.0   

     cash_proportion  Total Pay in million      % Cash  Previous Pay $  
2           0.554220                0.4491   55.421955    4.629897e+07  
3           0.027793                0.3598    2.779322    4.232941e+07  
4           0.139388                0.3336   13.938849    3.971429e+07  
5           0.487118                0.2484   48.711755    2.587500e+07  
6           0.195554                0.2204   19.555354    2.592941e+07  
..               ...                   ...         ...             ...  
96          0.980392                0.0153   98.039216    4.500000e+06  
97          0.989362                0.0094   98.936170    1.056180e+06  
98          0.876543                0.0081   87.654321    8.709677e+05  
100         1.000000                0.0004  100.000000    3.773585e+04  
101              NaN                0.0000         NaN    0.000000e+00  

[81 rows x 15 columns]

---

# **Question 3.6**
What was the average pay of these CEOs in 2014?  Does it make sense to compare this number to the number you computed in question 3?

In [168]:
average_pay_2014 = with_previous_compensation["Previous Pay $"].mean()
average_pay_2014

11649176.115603432

It has decreased a slight amount which makes sense because with the dataset most CEO's percents are negative

---

# **Question 3.7**

A skeptical student asks:

> "I already knew lots of ways to operate on each element of an array at once.  For example, I can multiply each element of `some_array` by 100 by writing `100*some_array`.  What good is `apply`?

How would you answer?  Discuss with a neighbor and place your answer below.

The apply function calls a function on each element in a column on a table, and apply calls the function we give it so we jsut have to write the name as an argument 

# Practice with histograms

Earlier, we computed the average pay among the CEOs in our 102-CEO dataset.  The average doesn't tell us everything about the amounts CEOs are paid, though.  Maybe just a few CEOs make the bulk of the money, even among these 102.

To learn more, let's use a histogram to display more information about the pay of the CEOs in our data set.

- - - - - - - - - - - - - - - - - - - - - - - - - - -

# **Question 4.1**

Make a histogram of the pay of the CEOs in `compensation`.  Be sure to:

* Give a title and label the axes
* Show the percents (not frequences) on the y axis
* Specifiy your own bins so that it is clear where each bin starts and each one ends

In [ ]:
n = len(compensation)
multiplier = np.ones(n)*100/n
multiplier

In [176]:
plt.hist(compensation["Total Pay $"]/10000000, edgecolor = 'black' , bins = np.arange(0,50,5), weights = multiplier)
plt.xticks(np.arange(0,55,5))
plt.title('CEO Total Pay')
plt.xlabel('Total Pay In Millions')
plt.ylabel('percentage')


TypeError: 'numpy.ndarray' object is not callable

- - - - - - - - - - - - - - - - - - - - - - - - - - -

# **Question 4.2**

Use the output of the ``plt.hist`` function to answer: How many CEOs made more than \$30 million?

Check out the cheatsheet if you don't remember how!

In [174]:
num_ceos_more_than_30_million = compensation[compensation['Total Pay $'] > 2999999].count()
num_ceos_more_than_30_million

Rank                                               90
Name                                               90
Company (Headquarters)                             90
Total Pay                                          90
% Change                                           90
Cash Pay                                           90
Equity Pay                                         90
Other Pay                                          90
Ratio of CEO pay to average industry worker pay    90
Total Pay $                                        90
Cash $                                             90
cash_proportion                                    90
Total Pay in million                               90
% Cash                                             90
dtype: int64


- - - - - - - - - - - - - - - - - - - - - - - - - - -

# **Question 4.3**

Do most CEOs make around the same amount, or are there some who make a lot more than the rest?  Discuss with someone near you.

There are three CEO's that make alot more then the rest

----